src/Nethermind/Nethermind.Merge.Plugin/Synchronization/PeerRefresher.cs

```diff
@@ -17,61 +17,202 @@
+            if (_logger.IsError) _logger.Error($"Exception in peer refresh. This is unexpected. {syncPeer}", exception);
+        }
+        finally
+        {
+            delaySource.Cancel();
+        }
+    }
+
+    internal async Task RefreshPeerForFcu(
+        ISyncPeer syncPeer,
+        Keccak headBlockhash,
+        Keccak headParentBlockhash,
+        Keccak finalizedBlockhash,
+        Task? delayTask,
+        CancellationToken token
+    ) {
+        
+        // headBlockhash is obtained together with headParentBlockhash
+        Task<BlockHeader[]> getHeadParentHeaderTask = syncPeer.GetBlockHeaders(headParentBlockhash, 2, 0, token);
+        Task<BlockHeader?> getFinalizedHeaderTask = finalizedBlockhash == Keccak.Zero 
+            ? Task.FromResult<BlockHeader?>(null)
+            : syncPeer.GetHeadBlockHeader(finalizedBlockhash, token);
+
+        Task getHeaderTask = Task.WhenAll(getFinalizedHeaderTask, getFinalizedHeaderTask);
+        Task firstToComplete = await Task.WhenAny(getHeaderTask, delayTask);
+
+        if (firstToComplete == delayTask)
+        {
+            _syncPeerPool.ReportRefreshFailed(syncPeer, "timeout");
+            return;
+        }
+
+        BlockHeader? headBlockHeader = null;
+        BlockHeader? headParentBlockHeader = null;
+        BlockHeader? finalizedBlockHeader = null;
+        try
+        {
+            BlockHeader[] headAndParentHeaders = await getHeadParentHeaderTask;
+            if (headAndParentHeaders.Length == 1 && headAndParentHeaders[0].Hash == headParentBlockhash)
+            {
+                headParentBlockHeader = headAndParentHeaders[0];
+            }
+            else if (headAndParentHeaders.Length == 2)
+            {
+                // Maybe the head is not the same as we expected. In that case, leave it as null
+                if (headBlockhash == headAndParentHeaders[1].Hash)
+                {
+                    headBlockHeader = headAndParentHeaders[1];
+                }
+                headParentBlockHeader = headAndParentHeaders[0];
+            }
+            else if (headAndParentHeaders.Length > 0)
+            {
+                if (_logger.IsTrace) _logger.Trace($"PeerRefreshForFCU unexpected response length when fetching header: {headAndParentHeaders.Length}");
+            }
+
+            finalizedBlockHeader = await getFinalizedHeaderTask;
+        }
+        catch (TaskCanceledException exception)
+        {
+            if (_logger.IsTrace)
+                _logger.Trace(
+                    $"PeerRefreshForFCU canceled for node: {syncPeer.Node:c}{Environment.NewLine}{exception}");
+            _syncPeerPool.ReportRefreshCancelled(syncPeer);
+            throw;
+        }
+        catch (Exception exception)
+        {
+            if (_logger.IsTrace)
+                _logger.Trace($"PeerRefreshForFCU failed for node: {syncPeer.Node:c}{Environment.NewLine}{exception}");
+            _syncPeerPool.ReportRefreshFailed(syncPeer, "faulted");
```

Comment:
Maybe pass exception to ReportRefreshFailed and improve reason a bit? Now we have 2x trace logs.